### PR TITLE
Update Installer

### PR DESCRIPTION
**Pull request by WinterSilence**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  |

### CODE REVIEW

File: Installer.php [hunk 1]
```diff
@@ -132,46 +132,49 @@ protected function addPackage(PackageInterface $package)
         $this->saveExtensions($extensions);
     }
 
-    protected function generateDefaultAlias(PackageInterface $package)
```
Comment (bizley) [inline Installer.php:135]: Why are you removing protected method? This is BC break.

Comment (WinterSilence) [reply Installer.php:135]: @bizley change name to correct, it's protected method + I'm don't find any packages extends that `Installer`

Comment (bizley) [reply Installer.php:135]: It doesn't matter that there are no public packages that extend it. Only private methods can be changed freely like that.

**Comment by WinterSilence** [comment]
@samdark it's refactoring, no changes in public api

File: Installer.php [hunk 2]
```diff
@@ -191,49 +190,26 @@ protected function loadExtensions()
...
-            mkdir(dirname($file), 0777, true);
-        }
-        $array = str_replace("'<vendor-dir>", '$vendorDir . \'', var_export($extensions, true));
-        file_put_contents($file, "<?php\n\n\$vendorDir = dirname(__DIR__);\n\nreturn $array;\n");
+        $this->filesystem->ensureDirectoryExists(dirname($file));
+        $array = str_replace("'$this->vendorDir/", '$vendorDir . \'/', var_export($extensions, true));
         // invalidate opcache of extensions.php if exists
         if (function_exists('opcache_invalidate')) {
-            @opcache_invalidate($file, true);
+            opcache_invalidate($file, true);
```
Comment (samdark) [inline Installer.php:204]: That would make Installer fail if extensions.php does not exist.

File: Installer.php [hunk 3]
```diff
@@ -345,12 +323,14 @@ public static function generateCookieValidationKey()
...
+            return bin2hex(random_bytes(16));
+        }
+
+        if (!function_exists('openssl_random_pseudo_bytes')) {
             throw new \Exception('The OpenSSL PHP extension is required by Yii2.');
         }
-        $length = 32;
-        $bytes = openssl_random_pseudo_bytes($length);
-        return strtr(substr(base64_encode($bytes), 0, $length), '+/=', '_-.');
+        return bin2hex(openssl_random_pseudo_bytes(16));
```
Comment (samdark) [inline Installer.php:333]: That doesn't seem to produce similar result.

File: Installer.php [hunk 4]
```diff
@@ -326,12 +301,15 @@ public static function setPermission(array $paths)
...
-     * Generates a cookie validation key for every app config listed in "config" in extra section.
+     * Generates a cookie validation key for every application configuration in "extra" section.
      * You can provide one or multiple parameters as the configuration files which need to have validation key inserted.
      */
     public static function generateCookieValidationKey()
     {
         $configs = func_get_args();
+        if (count($configs) === 1 && is_array($configs[0])) {
+            $configs = $configs[0];
+        }
```
Comment (samdark) [inline Installer.php:312]: What is it for?

File: Installer.php (hunk 2 above)
Comment (WinterSilence) [reply Installer.php:204]: @samdark `opcache_invalidate()` don't generate error

**Review by samdark — CHANGES_REQUESTED** [review]

File: Installer.php (hunk 4 above)
Comment (WinterSilence) [reply Installer.php:312]: @samdark to support configuration in format same to `setPermission` - uniform configuration form for all methods

File: Installer.php (hunk 3 above)
Comment (WinterSilence) [reply Installer.php:333]: @samdark sure, `base64_encode()` is not the same as `base64_encode()`, but it's faster and have same format `a-zA-Z0-9`

**Pull request closed without merging**